pkg/util/docker/event_stream.go

```diff
@@ -0,0 +1,115 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2017 Datadog, Inc.
+
+// +build docker
+
+package docker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	log "github.com/cihub/seelog"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+)
+
+//// Can't be unit tested, covered by the listeners/docker
+//// and dogstatsd/origin_detection integration tests.
+
+// eventSuscriber holds the state for a subscriber
+type eventSuscriber struct {
+	name       string
+	eventChan  chan *ContainerEvent
+	errorChan  chan error
+	cancelChan chan struct{}
+}
+
+// SubscribeToContainerEvents allows a package to suscribe to events from the event stream.
+// A unique subscriber name should be provided.
+func (d *DockerUtil) SubscribeToContainerEvents(name string) (<-chan *ContainerEvent, <-chan error, error) {
+	d.RLock()
+	if _, found := d.eventSuscribers[name]; found {
+		d.RUnlock()
+		return nil, nil, errors.New("already suscribed")
+	}
+	d.RUnlock()
+
+	sub := &eventSuscriber{
+		name:       name,
+		eventChan:  make(chan *ContainerEvent, 5),
+		errorChan:  make(chan error, 5),
+		cancelChan: make(chan struct{}),
+	}
+	d.Lock()
+	d.eventSuscribers[name] = sub
+	d.Unlock()
+
+	go d.streamEvents(sub.eventChan, sub.cancelChan)
```

Comment:
I'm not sure to understand the logic here, you spin off a new event listener routine each time there's a subscriber?

Why don't we have only one routine listening to events that loops over subscribers channels to dispatch them? 🤔

Reply:
That's the fan-out logic (sending the events to several subscriber channels from a single source channel) that was the bulk of the previous PR, but we decided to split in two.

That fan-out logic is not trivial as you need a channel per suscriber, and channel writes on full channels are blocking.

Reply:
Don't you already create 2 channels per subscriber here?